src/main/scala/calc/Compiler.scala

```diff
@@ -66,17 +73,129 @@ object Compiler {
    *
    *  This is the main method you have to implement.
    */
-  def compileExpr(tree: Tree): irt.Tree = {
+  def compileExpr(tree: Tree, scope: Scope=Map.empty[String, TreeType]): irt.Tree = {
     // TODO
     implicit val pos = tree.pos
 
     tree match {
-      case Literal(value) =>
+      case Literal(value, _) =>
         irt.DoubleLiteral(value)
 
+      case i@Ident(name, tp) => scope.get(name) match {
+        // we have to cast parameters to Double, as they are always of type AnyType:
+        case Some(t) if tp.isParam =>
+          irt.Unbox(irt.VarRef(irt.Ident(name))(irtpe.AnyType), 'D')
+
+        case Some(t) => irt.VarRef(irt.Ident(name))(tp)
+
+        // should never happen, typechecker will spot it first:
+        case None => throw new Exception(s"Not in scope: $name")
+      }
+
+      case BinaryOp(op, lhs, rhs, _) =>
+        irt.BinaryOp(parseBinOp(op),
+            compileExpr(lhs, scope), compileExpr(rhs, scope))
+
+      case Let(ident, value, body, _) =>
+        val newScope = scope + (ident.name -> value.tp)
+        val valueC = compileExpr(value, scope)
+        val bodyC = compileExpr(body, newScope)
+        val varDefC = irt.VarDef(irt.Ident(ident.name),
+            ident.tp, mutable = false, valueC)
+
+        irt.Block(List(varDefC, bodyC))
+
+      case If(cond, thenp, elsep, tp) =>
+        // in this language, we could just optimize-out
+        // the if statement (condition is always a number)
+        val condC  = evalIfCond(cond)
+        val thenpC = compileExpr(thenp, scope)
+        val elsepC = compileExpr(elsep, scope)
+
+        irt.If(condC, thenpC, elsepC)(tp)
+
+      case Closure(params, body, _) =>
+        val captureParamsC = paramsFromScope(scope)
```

Comment:
All technically correct, this is inefficient. You'll capture every single variable available in the scope of the closure. Instead, you should only use those variables that are actually *used* in the body of the closure. This is a set known as the "free variables" of the body, if you want to Google for it.